File: prep/implementation/patterns_reference_implementation/subsets/Subsets.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(2^n) / O(2^n)
public class Subsets {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = subsets(a);
        for (var sub : res) {
          for (int e : sub) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }

  private static List<List<Integer>> subsets(int[] a) {
    List<List<Integer>> allSubs = new ArrayList<>();
    allSubs.add(new ArrayList<>());

    for (int e : a) {
      int n = allSubs.size();
      for (int j = 0; j < n; j++) {
        List<Integer> currSub = new ArrayList<>(allSubs.get(j));
        currSub.add(e);
        allSubs.add(currSub);
      }
    }

    return allSubs;
  }
}

File: prep/implementation/patterns_reference_implementation/subsets/subsets.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// O(2^n) / O(2^n)
vector<vector<int>> subsets(const vector<int> &a) {
  vector<vector<int>> all_subs;
  all_subs.push_back({});

  for (int e : a) {
    int n = all_subs.size();
    for (int i = 0; i < n; i++) {
      vector<int> curr_sub = all_subs[i];
      curr_sub.push_back(e);
      all_subs.emplace_back(curr_sub);
    }
  }

  return all_subs;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = subsets(a);
    for (auto r : res) {
      for (int e : r) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}